pkg/storegateway/indexheader/bucket_reader.go

```diff
@@ -0,0 +1,48 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package indexheader
+
+import (
+	"context"
+	"errors"
+	"io"
+	"math"
+	"time"
+
+	"github.com/thanos-io/objstore"
+)
+
+var errLowThroughput = errors.New("throughput is lower than expected minimum")
+
+type bucketRangeReader interface {
+	// GetRange returns a new range reader for the given object name and range.
+	GetRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error)
+	// Attributes returns information about the specified object.
+	Attributes(ctx context.Context, name string) (objstore.ObjectAttributes, error)
+}
+
+type minThroughputBucketRangeReader struct {
+	bucketRangeReader
+	// Minimum expected throughput in bytes/s.
+	throughput int
+}
+
+func bucketRangeReaderWithEnforcedThroughput(bkt bucketRangeReader, throughput int) bucketRangeReader {
+	return &minThroughputBucketRangeReader{
+		bucketRangeReader: bkt,
+		throughput:        throughput,
+	}
+}
+
+func (m *minThroughputBucketRangeReader) GetRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error) {
+	ctx, cancel := context.WithTimeoutCause(ctx, m.expectedTime(length), errLowThroughput)
+	defer cancel()
```

Comment:
don't we need the context to survive beyond the scope of this method? The io.ReadCloser I think uses the `ctx` we build